pkg/internalapis/cnsvolumeinfo/cnsvolumeinfoservice.go

```diff
@@ -0,0 +1,193 @@
+	CRDGroupName = "cns.vmware.com"
+)
+
+// VolumeInfoService exposes interfaces to support Operate on cnsvolumeinfo CR
+// for multi vCenter CSI topology feature.
+// It will maintain internal state to map volume id to vCenter
+
+type VolumeInfoService interface {
+	// GetvCenterForVolumeID return vCenter for the given VolumeID
+	GetvCenterForVolumeID(ctx context.Context, volumeID string) (string, error)
+
+	// CreateVolumeInfo creates VolumeInfo CR to persist VolumeID to vCenter mapping
+	CreateVolumeInfo(ctx context.Context, volumeID string, vCenter string) error
+
+	// DeleteVolumeInfo deletes VolumeInfo CR for the given VolumeID
+	DeleteVolumeInfo(ctx context.Context, volumeID string) error
+}
+
+// InitVolumeInfoService returns the singleton VolumeInfoService.
+func InitVolumeInfoService(ctx context.Context) (VolumeInfoService, error) {
+	log := logger.GetLogger(ctx)
+	if volumeInfoServiceInstance == nil {
+		log.Info("Initializing volumeInfo service...")
+		// This is idempotent if CRD is pre-created then we continue with
+		// initialization of volumeInfoServiceInstance.
+		volumeInfoServiceInitErr := k8s.CreateCustomResourceDefinitionFromManifest(ctx,
+			cnsvolumeinfoconfig.EmbedCnsVolumeInfoFile, cnsvolumeinfoconfig.EmbedCnsVolumeInfoFileName)
+
+		if volumeInfoServiceInitErr != nil {
+			log.Errorf("failed to create volume info CRD. Error: %v", volumeInfoServiceInitErr)
+			return nil, volumeInfoServiceInitErr
+		}
+		config, volumeInfoServiceInitErr := k8s.GetKubeConfig(ctx)
+		if volumeInfoServiceInitErr != nil {
+			log.Errorf("failed to get kubeconfig. err: %v", volumeInfoServiceInitErr)
+			return nil, volumeInfoServiceInitErr
+		}
+		volumeInfoServiceInstance = &volumeInfo{
+			volumeIDTovCenter: sync.Map{},
+		}
+		volumeInfoServiceInstance.k8sClient, volumeInfoServiceInitErr =
+			k8s.NewClientForGroup(ctx, config, CRDGroupName)
+		if volumeInfoServiceInitErr != nil {
+			volumeInfoServiceInstance = nil
+			log.Errorf("failed to create k8sClient. Err: %v", volumeInfoServiceInitErr)
+			return nil, volumeInfoServiceInitErr
+		}
+		go func() {
+			log.Debugf("Starting Informer for cnsvolumeinfo")
+			informer, err := k8s.GetDynamicInformer(ctx, cnsvolumeinfov1alpha1.SchemeGroupVersion.Group,
+				cnsvolumeinfov1alpha1.SchemeGroupVersion.Version, "cnsvolumeinfoes",
+				CSINamespace, config, true)
+			if err != nil {
+				log.Errorf("failed to create dynamic informer for cnsvolumeinfoes CRD. Err: %v", err)
+			}
+			handlers := cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					log.Debugf("received add event for cnsvolumeinfo CR!")
+					var volumeInfoObject cnsvolumeinfov1alpha1.CNSVolumeInfo
+					err := runtime.DefaultUnstructuredConverter.FromUnstructured(
+						obj.(*unstructured.Unstructured).Object, &volumeInfoObject)
+					if err != nil {
+						log.Errorf("failed to cast object to volumeInfoObject. err: %v", err)
+						return
+					}
+					volumeInfoServiceInstance.volumeIDTovCenter.Store(
```

Comment:
Can we not use the informer cache itself? Do we need a separate map for it?

Reply:
used the separate map to get needed mapping very quickly using the volumeID as key.
Can we achieve the same using cache? We don't want to use the loop to get the requested mapping.

Reply:
Yes we can achieve the same using the informer cache itself. You can use the `GetByKey` function of the Store like we did in https://github.com/kubernetes-sigs/vsphere-csi-driver/blob/master/pkg/csi/service/common/commonco/k8sorchestrator/topology.go#L1215

Reply:
done. addressed.